File: backend/internal/httpapi/recipes.go
package httpapi

import (
	"context"
	"errors"
	"net/http"
	"strings"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"
	"github.com/saiaj/cooking_app/backend/internal/httpapi/response"
)

type recipeTagResponse struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type recipeIngredientResponse struct {
	ID           string       `json:"id"`
	Position     int          `json:"position"`
	Quantity     *float64     `json:"quantity"`
	QuantityText *string      `json:"quantity_text"`
	Unit         *string      `json:"unit"`
	Item         itemResponse `json:"item"`
	Prep         *string      `json:"prep"`
	Notes        *string      `json:"notes"`
	OriginalText *string      `json:"original_text"`
}

type recipeStepResponse struct {
	ID          string `json:"id"`
	StepNumber  int    `json:"step_number"`
	Instruction string `json:"instruction"`
}

type recipeDetailResponse struct {
	ID               string                     `json:"id"`
	Title            string                     `json:"title"`
	Servings         int32                      `json:"servings"`
	PrepTimeMinutes  int32                      `json:"prep_time_minutes"`
	TotalTimeMinutes int32                      `json:"total_time_minutes"`
	SourceURL        *string                    `json:"source_url"`
	Notes            *string                    `json:"notes"`
	RecipeBookID     *string                    `json:"recipe_book_id"`
	Tags             []recipeTagResponse        `json:"tags"`
	Ingredients      []recipeIngredientResponse `json:"ingredients"`
	Steps            []recipeStepResponse       `json:"steps"`
	CreatedAt        string                     `json:"created_at"`
	CreatedBy        string                     `json:"created_by"`
	UpdatedAt        string                     `json:"updated_at"`
	UpdatedBy        string                     `json:"updated_by"`
	DeletedAt        *string                    `json:"deleted_at"`
}

func (a *App) handleRecipesCreate(w http.ResponseWriter, r *http.Request) error {
	info, ok := authInfoFromRequest(r)
	if !ok {
		return errUnauthorized("unauthorized")
	}

	var req createRecipeRequest
	if err := a.decodeJSON(w, r, &req); err != nil {
		return err
	}

	if errs := validateCreateRecipeRequest(req); len(errs) > 0 {
		return errValidation(errs)
	}

	ctx := r.Context()
	userID := pgtype.UUID{Bytes: info.UserID, Valid: true}

	recipeID, err := createRecipeUsecase(ctx, a.recipeWorkflows(), userID, req)
	if err != nil {
		return mapRecipeUsecaseError(err)
	}

	detail, err := a.loadRecipeDetail(ctx, recipeID)
	if err != nil {
		return errInternal(err)
	}

	if err := response.WriteJSON(w, http.StatusCreated, detail); err != nil {
		a.logger.Warn("write failed", "err", err, "path", "/api/v1/recipes")
	}
	return nil
}

func (a *App) handleRecipesGet(w http.ResponseWriter, r *http.Request) error {
	if _, ok := authInfoFromRequest(r); !ok {
		return errUnauthorized("unauthorized")
	}

	id, err := parseUUIDParam(r, "id")
	if err != nil {
		return err
	}

	detail, err := a.loadRecipeDetail(r.Context(), pgtype.UUID{Bytes: id, Valid: true})
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return errNotFound()
		}
		return errInternal(err)
	}

	if err := response.WriteJSON(w, http.StatusOK, detail); err != nil {
		a.logger.Warn("write failed", "err", err, "path", "/api/v1/recipes/{id}")
	}
	return nil
}

func (a *App) loadRecipeDetail(ctx context.Context, id pgtype.UUID) (recipeDetailResponse, error) {
	row, err := a.queries.GetRecipeByID(ctx, id)
	if err != nil {
		return recipeDetailResponse{}, err
	}

	ingredients, err := a.queries.ListRecipeIngredientsByRecipeID(ctx, id)
	if err != nil {
		return recipeDetailResponse{}, err
	}
	steps, err := a.queries.ListRecipeStepsByRecipeID(ctx, id)
	if err != nil {
		return recipeDetailResponse{}, err
	}
	tags, err := a.queries.ListRecipeTagsByRecipeID(ctx, id)
	if err != nil {
		return recipeDetailResponse{}, err
	}

	outIngredients := make([]recipeIngredientResponse, 0, len(ingredients))
	for _, ing := range ingredients {
		var quantity *float64
		if ing.Quantity.Valid {
			f8, err := ing.Quantity.Float64Value()
			if err != nil {
				return recipeDetailResponse{}, err
			}
			if f8.Valid {
				q := f8.Float64
				quantity = &q
			}
		}
		outIngredients = append(outIngredients, recipeIngredientResponse{
			ID:           uuidString(ing.ID),
			Position:     int(ing.Position),
			Quantity:     quantity,
			QuantityText: textStringPtr(ing.QuantityText),
			Unit:         textStringPtr(ing.Unit),
			Item: itemResponse{
				ID:       uuidString(ing.ItemID),
				Name:     ing.ItemName,
				StoreURL: textStringPtr(ing.ItemStoreUrl),
				Aisle: buildAisleResponse(
					ing.ItemAisleID,
					ing.AisleName,
					ing.AisleSortGroup,
					ing.AisleSortOrder,
					ing.AisleNumericValue,
				),
			},
			Prep:         textStringPtr(ing.Prep),
			Notes:        textStringPtr(ing.Notes),
			OriginalText: textStringPtr(ing.OriginalText),
		})
	}

	outSteps := make([]recipeStepResponse, 0, len(steps))
	for _, s := range steps {
		outSteps = append(outSteps, recipeStepResponse{
			ID:          uuidString(s.ID),
			StepNumber:  int(s.StepNumber),
			Instruction: s.Instruction,
		})
	}

	outTags := make([]recipeTagResponse, 0, len(tags))
	for _, t := range tags {
		outTags = append(outTags, recipeTagResponse{ID: uuidString(t.ID), Name: t.Name})
	}

	return recipeDetailResponse{
		ID:               uuidString(row.ID),
		Title:            row.Title,
		Servings:         row.Servings,
		PrepTimeMinutes:  row.PrepTimeMinutes,
		TotalTimeMinutes: row.TotalTimeMinutes,
		SourceURL:        textStringPtr(row.SourceUrl),
		Notes:            textStringPtr(row.Notes),
		RecipeBookID:     uuidStringPtr(row.RecipeBookID),
		Tags:             outTags,
		Ingredients:      outIngredients,
		Steps:            outSteps,
		CreatedAt:        timeString(row.CreatedAt),
		CreatedBy:        uuidString(row.CreatedBy),
		UpdatedAt:        timeString(row.UpdatedAt),
		UpdatedBy:        uuidString(row.UpdatedBy),
		DeletedAt:        timeStringPtr(row.DeletedAt),
	}, nil
}

func textPtrToPG(value *string) pgtype.Text {
	if value == nil {
		return pgtype.Text{}
	}
	trimmed := strings.TrimSpace(*value)
	if trimmed == "" {
		return pgtype.Text{}
	}
	return pgtype.Text{String: trimmed, Valid: true}
}

func textStringPtr(v pgtype.Text) *string {
	if !v.Valid {
		return nil
	}
	s := v.String
	return &s
}

func uuidStringPtr(v pgtype.UUID) *string {
	if !v.Valid {
		return nil
	}
	s := uuid.UUID(v.Bytes).String()
	return &s
}

func numericPtrFromFloat64(v *float64) (pgtype.Numeric, error) {
	if v == nil {
		return pgtype.Numeric{}, nil
	}
	var n pgtype.Numeric
	if err := n.Scan(*v); err != nil {
		return pgtype.Numeric{}, err
	}
	return n, nil
}

func uuidPtrToPG(id *string) (pgtype.UUID, error) {
	if id == nil {
		return pgtype.UUID{}, nil
	}
	trimmed := strings.TrimSpace(*id)
	if trimmed == "" {
		return pgtype.UUID{}, nil
	}
	parsed, err := uuid.Parse(trimmed)
	if err != nil {
		return pgtype.UUID{}, err
	}
	return pgtype.UUID{Bytes: parsed, Valid: true}, nil
}

func uuidsToPG(ids []string) ([]pgtype.UUID, error) {
	out := make([]pgtype.UUID, 0, len(ids))
	for _, id := range ids {
		id = strings.TrimSpace(id)
		if id == "" {
			continue
		}
		parsed, err := uuid.Parse(id)
		if err != nil {
			return nil, err
		}
		out = append(out, pgtype.UUID{Bytes: parsed, Valid: true})
	}
	return out, nil
}

func timeStringPtr(v pgtype.Timestamptz) *string {
	if !v.Valid {
		return nil
	}
	s := timeString(v)
	return &s
}

func intToInt32Checked(v int) (int32, bool) {
	if v > maxInt32 {
		return 0, false
	}
	minInt32 := -maxInt32 - 1
	if v < minInt32 {
		return 0, false
	}
	return int32(v), true //nolint:gosec // bounds checked above
}

File: frontend/src/api/recipes.ts
import { apiFetchJSON } from './client'
import type { Item } from './items'

export type RecipeTag = {
  id: string
  name: string
}

export type RecipeListItem = {
  id: string
  title: string
  servings: number
  prep_time_minutes: number
  total_time_minutes: number
  source_url: string | null
  notes: string | null
  recipe_book_id: string | null
  tags: RecipeTag[]
  deleted_at: string | null
  updated_at: string
}

export type RecipeListResponse = {
  items: RecipeListItem[]
  next_cursor: string | null
}

export type RecipeIngredient = {
  id: string
  position: number
  quantity: number | null
  quantity_text: string | null
  unit: string | null
  item: Item
  prep: string | null
  notes: string | null
  original_text: string | null
}

export type RecipeStep = {
  id: string
  step_number: number
  instruction: string
}

export type RecipeIngredientUpsert = {
  position: number
  quantity: number | null
  quantity_text: string | null
  unit: string | null
  item_id: string | null
  item_name: string | null
  prep: string | null
  notes: string | null
  original_text: string | null
}

export type RecipeStepUpsert = {
  step_number: number
  instruction: string
}

export type RecipeUpsertRequest = {
  title: string
  servings: number
  prep_time_minutes: number
  total_time_minutes: number
  source_url: string | null
  notes: string | null
  recipe_book_id: string | null
  tag_ids: string[]
  ingredients: RecipeIngredientUpsert[]
  steps: RecipeStepUpsert[]
}

export type RecipeDetail = {
  id: string
  title: string
  servings: number
  prep_time_minutes: number
  total_time_minutes: number
  source_url: string | null
  notes: string | null
  recipe_book_id: string | null
  tags: RecipeTag[]
  ingredients: RecipeIngredient[]
  steps: RecipeStep[]
  created_at: string
  created_by: string
  updated_at: string
  updated_by: string
  deleted_at: string | null
}

export type ListRecipesParams = {
  q?: string
  book_id?: string
  tag_id?: string
  include_deleted?: boolean
  limit?: number
  cursor?: string
}

export function listRecipes(
  params: ListRecipesParams = {},
): Promise<RecipeListResponse> {
  const qp = new URLSearchParams()

  if (params.q?.trim()) qp.set('q', params.q.trim())
  if (params.book_id) qp.set('book_id', params.book_id)
  if (params.tag_id) qp.set('tag_id', params.tag_id)
  if (params.include_deleted) qp.set('include_deleted', 'true')
  if (typeof params.limit === 'number') qp.set('limit', String(params.limit))
  if (params.cursor) qp.set('cursor', params.cursor)

  const url = qp.size ? `/api/v1/recipes?${qp.toString()}` : '/api/v1/recipes'
  return apiFetchJSON<RecipeListResponse>(url)
}

export function getRecipe(id: string): Promise<RecipeDetail> {
  return apiFetchJSON<RecipeDetail>(`/api/v1/recipes/${id}`)
}

export function createRecipe(
  params: RecipeUpsertRequest,
): Promise<RecipeDetail> {
  return apiFetchJSON<RecipeDetail>('/api/v1/recipes', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
}

export function updateRecipe(
  id: string,
  params: RecipeUpsertRequest,
): Promise<RecipeDetail> {
  return apiFetchJSON<RecipeDetail>(`/api/v1/recipes/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
}

export function deleteRecipe(id: string): Promise<void> {
  return apiFetchJSON<void>(`/api/v1/recipes/${id}`, { method: 'DELETE' })
}

export function restoreRecipe(id: string): Promise<void> {
  return apiFetchJSON<void>(`/api/v1/recipes/${id}/restore`, { method: 'PUT' })
}
